### ISSUE TITLE
Occasional SQL errors when writting to Options

### ISSUE THREAD

**Issue by patrickebates** [ProjectNami/projectnami#140]
Seeing inconsistent SQL errors when writing to wp_options.  Here's what we know so far...

- So far only seen internally when writing transients
- Values are very large, measuring in hundreds of KBs
- Values are serialized objects
- Values contain XML strings, such as the WP news feed

Haven't seen anything in the base SQL statement that leads us to believe it is simply a syntax error in the code, or all attempts to write to wp_options would be failing.  Here's a list of other possible causes...

- Issue with SQL escaping of the string
- Data is being truncated somewhere (so far all of the error messages I have seen contain incomplete SQL statements)
- Order of operations issue between serialization and SQL escaping

Use of caching plugins will prevent nearly all instances of these errors, as transients will no longer be written to wp_options.

Looking for more information and error logs from our users.  Thanks.

**Comment by patrickebates** [comment]
Thanks to the very small error reported in #121 I have been able to find a consistent pattern to all of the errors.  The SQL statements all have issues at Requests_Utility_CaseInsensitiveDictionary regardless of the length of the statement.

**Comment by dulfe** [comment]
I found this issue when using **Gravity Forms**... after activating the plugin it tries to get some update information and save it as *transient*, it fails.

Log files: 
[pn_logs.zip](https://github.com/ProjectNami/projectnami/files/525282/pn_logs.zip)

I also noticed that even without GF, the wp-admin dashboard fails when trying to get the list of **news** (with ` define('WP_DEBUG', true)` )... I think it has something to do with quotes in the value been saved in the wp_options.... mixed with the logic that serializes the value been saved.

I will try and get a sample that can be used for testing.

Daniel.

**Comment by patrickebates** [comment]
Yes, the problem with transients and news are what we first identified.  But I didn't have an example outside of them until the error you reported in the other thread.  That helped me identify a common failure point with the serialized string for Requests_Utility_CaseInsensitiveDictionary.  So far, I've not been able to identify the exact issue with the text.  But once we locate it, if the issue is outside of WP/PN, perhaps in the DB connection layer, I have contacts who can get to the bottom of it.

**Comment by patrickebates** [comment]
Found it.  Requests_Utility_CaseInsensitiveDictionary has a Protected variable.

http://php.net/manual/en/function.serialize.php

Note: 
Object's private members have the class name prepended to the member name; protected members have a '*' prepended to the member name. These prepended values have null bytes on either side. 

The null bytes can't be passed to SQL in a string.  Looking at options, including possibly changing the scope of the variable.

**Comment by dulfe** [comment]
@patrickebates I was thinking about this for the last few days and these are my two (well, three) cents:

- Trying to add support for `NULL` values in the Translation layer using strings seems too complicated. I just could not find any information about how to encode null values... so I think it will require too many changes to code that is already working "fine". (I could be way wrong here)... on the other hand, that would work with ANY class using protected values and any tables.
- Changing the scope to `public` is easy, but that would only work on that specific class and if the user tries to encode another class with protected values it will break again. And to be honest, that property was defined as protected for a reason.
- Changing how `add_option`, `get_option` and `update_option` code to handle those null values; That would work with any class that has protected values but ONLY in wp_options.

Of the 3, the last one seems quick and easy and it provides a good balance between flexibility and effort :)

As I proof of concept I did:

```php
function add_option( $option, $value = '', $deprecated = '', $autoload = 'yes' ) {
	global $wpdb;

	if ( !empty( $deprecated ) )
		_deprecated_argument( __FUNCTION__, '2.3.0' );

	$option = trim($option);
	if ( empty($option) )
		return false;

	wp_protect_special_option( $option );

	if ( is_object($value) )
		$value = clone $value;

	$value = sanitize_option( $option, $value );

	// Make sure the option doesn't already exist. We can check the 'notoptions' cache before we ask for a db query
	$notoptions = wp_cache_get( 'notoptions', 'options' );
	if ( !is_array( $notoptions ) || !isset( $notoptions[$option] ) )
		/** This filter is documented in wp-includes/option.php */
		if ( apply_filters( 'default_option_' . $option, false, $option ) !== get_option( $option ) )
			return false;

	$serialized_value = maybe_serialize( $value );
    // Remove NULL markers for protected values.
    $serialized_value = preg_replace('/(s:\d+:\")(\0\*\0)(\w+\")/', '$1*$3', $serialized_value);
	$autoload = ( 'no' === $autoload || false === $autoload ) ? 'no' : 'yes';

	/**
	 * Fires before an option is added.
	 *
	 * @since 2.9.0
	 *
	 * @param string $option Name of the option to add.
	 * @param mixed  $value  Value of the option.
	 */
	do_action( 'add_option', $option, $value );

	$result = $wpdb->query( $wpdb->prepare( "IF NOT EXISTS (SELECT * FROM [$wpdb->options] WHERE [option_name] = '%s') INSERT INTO [$wpdb->options] ([option_name], [option_value], [autoload]) VALUES ('%s', '%s', '%s') else UPDATE [$wpdb->options] set [option_value] = '%s', [autoload] = '%s' where [option_name] = '%s'", array( $option, $option, $serialized_value, $autoload, $serialized_value, $autoload, $option ) ) );
	if ( $result === false )
		return false;

	if ( ! wp_installing() ) {
		if ( 'yes' == $autoload ) {
			$alloptions = wp_load_alloptions();
			$alloptions[ $option ] = $serialized_value;
			wp_cache_set( 'alloptions', $alloptions, 'options' );
		} else {
			wp_cache_set( $option, $serialized_value, 'options' );
		}
	}

	// This option exists now
	$notoptions = wp_cache_get( 'notoptions', 'options' ); // yes, again... we need it to be fresh
	if ( is_array( $notoptions ) && isset( $notoptions[$option] ) ) {
		unset( $notoptions[$option] );
		wp_cache_set( 'notoptions', $notoptions, 'options' );
	}

	/**
	 * Fires after a specific option has been added.
	 *
	 * The dynamic portion of the hook name, `$option`, refers to the option name.
	 *
	 * @since 2.5.0 As "add_option_{$name}"
	 * @since 3.0.0
	 *
	 * @param string $option Name of the option to add.
	 * @param mixed  $value  Value of the option.
	 */
	do_action( "add_option_{$option}", $option, $value );

	/**
	 * Fires after an option has been added.
	 *
	 * @since 2.9.0
	 *
	 * @param string $option Name of the added option.
	 * @param mixed  $value  Value of the option.
	 */
	do_action( 'added_option', $option, $value );
	return true;
}

function get_option( $option, $default = false ) {
	global $wpdb;

	$option = trim( $option );
	if ( empty( $option ) )
		return false;

	/**
	 * Filters the value of an existing option before it is retrieved.
	 *
	 * The dynamic portion of the hook name, `$option`, refers to the option name.
	 *
	 * Passing a truthy value to the filter will short-circuit retrieving
	 * the option value, returning the passed value instead.
	 *
	 * @since 1.5.0
	 * @since 4.4.0 The `$option` parameter was added.
	 *
	 * @param bool|mixed $pre_option Value to return instead of the option value.
	 *                               Default false to skip it.
	 * @param string     $option     Option name.
	 */
	$pre = apply_filters( 'pre_option_' . $option, false, $option );
	if ( false !== $pre )
		return $pre;

	if ( defined( 'WP_SETUP_CONFIG' ) )
		return false;

	if ( ! wp_installing() ) {
		// prevent non-existent options from triggering multiple queries
		$notoptions = wp_cache_get( 'notoptions', 'options' );
		if ( isset( $notoptions[ $option ] ) ) {
			/**
			 * Filters the default value for an option.
			 *
			 * The dynamic portion of the hook name, `$option`, refers to the option name.
			 *
			 * @since 3.4.0
			 * @since 4.4.0 The `$option` parameter was added.
			 *
			 * @param mixed  $default The default value to return if the option does not exist
			 *                        in the database.
			 * @param string $option  Option name.
			 */
			return apply_filters( 'default_option_' . $option, $default, $option );
		}

		$alloptions = wp_load_alloptions();

		if ( isset( $alloptions[$option] ) ) {
			$value = $alloptions[$option];
		} else {
			$value = wp_cache_get( $option, 'options' );

			if ( false === $value ) {
				$row = $wpdb->get_row( $wpdb->prepare( "SELECT TOP 1 option_value FROM $wpdb->options WHERE option_name = %s", $option ) );

				// Has to be get_row instead of get_var because of funkiness with 0, false, null values
				if ( is_object( $row ) ) {
					$value = $row->option_value;
					wp_cache_add( $option, $value, 'options' );
				} else { // option does not exist, so we must cache its non-existence
					if ( ! is_array( $notoptions ) ) {
						 $notoptions = array();
					}
					$notoptions[$option] = true;
					wp_cache_set( 'notoptions', $notoptions, 'options' );

					/** This filter is documented in wp-includes/option.php */
					return apply_filters( 'default_option_' . $option, $default, $option );
				}
			}
		}
	} else {
		$suppress = $wpdb->suppress_errors();
		$row = $wpdb->get_row( $wpdb->prepare( "SELECT TOP 1 option_value FROM $wpdb->options WHERE option_name = %s", $option ) );
		$wpdb->suppress_errors( $suppress );
		if ( is_object( $row ) ) {
			$value = $row->option_value;
		} else {
			/** This filter is documented in wp-includes/option.php */
			return apply_filters( 'default_option_' . $option, $default, $option );
		}
	}

	// If home is not set use siteurl.
	if ( 'home' == $option && '' == $value )
		return get_option( 'siteurl' );

	if ( in_array( $option, array('siteurl', 'home', 'category_base', 'tag_base') ) )
		$value = untrailingslashit( $value );

	/**
	 * Filters the value of an existing option.
	 *
	 * The dynamic portion of the hook name, `$option`, refers to the option name.
	 *
	 * @since 1.5.0 As 'option_' . $setting
	 * @since 3.0.0
	 * @since 4.4.0 The `$option` parameter was added.
	 *
	 * @param mixed  $value  Value of the option. If stored serialized, it will be
	 *                       unserialized prior to being returned.
	 * @param string $option Option name.
	 */
    // Insert NULL markers for protected values.
    $value = preg_replace('/(s:\d+:\")(\*)(\w+\")/', "$1\0*\0$3", $value);
	$unserialized_value = maybe_unserialize( $value );
	return apply_filters( 'option_' . $option, $unserialized_value, $option );
}

function update_option( $option, $value, $autoload = null ) {
	global $wpdb;

	$option = trim($option);
	if ( empty($option) )
		return false;

	wp_protect_special_option( $option );

	if ( is_object( $value ) )
		$value = clone $value;

	$value = sanitize_option( $option, $value );
	$old_value = get_option( $option );

	/**
	 * Filters a specific option before its value is (maybe) serialized and updated.
	 *
	 * The dynamic portion of the hook name, `$option`, refers to the option name.
	 *
	 * @since 2.6.0
	 * @since 4.4.0 The `$option` parameter was added.
	 *
	 * @param mixed  $value     The new, unserialized option value.
	 * @param mixed  $old_value The old option value.
	 * @param string $option    Option name.
	 */
	$value = apply_filters( 'pre_update_option_' . $option, $value, $old_value, $option );

	/**
	 * Filters an option before its value is (maybe) serialized and updated.
	 *
	 * @since 3.9.0
	 *
	 * @param mixed  $value     The new, unserialized option value.
	 * @param string $option    Name of the option.
	 * @param mixed  $old_value The old option value.
	 */
	$value = apply_filters( 'pre_update_option', $value, $option, $old_value );

	// If the new and old values are the same, no need to update.
	if ( $value === $old_value )
		return false;

	/** This filter is documented in wp-includes/option.php */
	if ( apply_filters( 'default_option_' . $option, false, $option ) === $old_value ) {
		// Default setting for new options is 'yes'.
		if ( null === $autoload ) {
			$autoload = 'yes';
		}

		return add_option( $option, $value, '', $autoload );
	}

	$serialized_value = maybe_serialize( $value );
    // Remove NULL markers for protected values.
    $serialized_value = preg_replace('/(s:\d+:\")(\0\*\0)(\w+\")/', '$1*$3', $serialized_value);

	/**
	 * Fires immediately before an option value is updated.
	 *
	 * @since 2.9.0
	 *
	 * @param string $option    Name of the option to update.
	 * @param mixed  $old_value The old option value.
	 * @param mixed  $value     The new option value.
	 */
	do_action( 'update_option', $option, $old_value, $value );

	$update_args = array(
		'option_value' => $serialized_value,
	);

	if ( null !== $autoload ) {
		$update_args['autoload'] = ( 'no' === $autoload || false === $autoload ) ? 'no' : 'yes';
	}

	$result = $wpdb->update( $wpdb->options, $update_args, array( 'option_name' => $option ) );
	if ( ! $result )
		return false;

	$notoptions = wp_cache_get( 'notoptions', 'options' );
	if ( is_array( $notoptions ) && isset( $notoptions[$option] ) ) {
		unset( $notoptions[$option] );
		wp_cache_set( 'notoptions', $notoptions, 'options' );
	}

	if ( ! wp_installing() ) {
		$alloptions = wp_load_alloptions();
		if ( isset( $alloptions[$option] ) ) {
			$alloptions[ $option ] = $serialized_value;
			wp_cache_set( 'alloptions', $alloptions, 'options' );
		} else {
			wp_cache_set( $option, $serialized_value, 'options' );
		}
	}

	/**
	 * Fires after the value of a specific option has been successfully updated.
	 *
	 * The dynamic portion of the hook name, `$option`, refers to the option name.
	 *
	 * @since 2.0.1
	 * @since 4.4.0 The `$option` parameter was added.
	 *
	 * @param mixed  $old_value The old option value.
	 * @param mixed  $value     The new option value.
	 * @param string $option    Option name.
	 */
	do_action( "update_option_{$option}", $old_value, $value, $option );

	/**
	 * Fires after the value of an option has been successfully updated.
	 *
	 * @since 2.9.0
	 *
	 * @param string $option    Name of the updated option.
	 * @param mixed  $old_value The old option value.
	 * @param mixed  $value     The new option value.
	 */
	do_action( 'updated_option', $option, $old_value, $value );
	return true;
}
```
I just added code to replace `NULL * NULL` with only `*` when saving and inserting those NULLs when reading. (Just 1 line in `add_option`, `update_option` and 2 lines in `get_option`)

It is not elegant nor perfect, but is a quick workaround without changing the translation layer.... of course changing the db/translation layer would be the best solution.

Daniel.

**Comment by patrickebates** [comment]
Had some emails with the guy at MS responsible for the PHP SqlSrv drivers.  Have a triage option, and some fix options.

Triage
- Change the variable in Requests_Utility_CaseInsensitiveDictionary to Public

Fix options
- Update maybe_serialize and maybe_unserialize to replace the null bytes with a unique string going in and replace them coming out
- Change the SQL call for inserting options to use paramaters (doing this to some queries is on my short-term To Do List for PN)

**Comment by patrickebates** [comment]
I've gone so far as to test this live on a production server that was having significant issues with this error.  Appears to have cleared everything up immediately.

Will be looking to release a version containing this change very soon.

**Comment by patrickebates** [comment]
Fixed in 1.4.3
